src/Entities/Player.cpp

```diff
@@ -339,6 +314,79 @@ void cPlayer::Tick(std::chrono::milliseconds a_Dt, cChunk & a_Chunk)
 
 
 
+void cPlayer::TickFreezeCode(bool a_MyChunkIsSent)
+{
+	// This function is ticked by the player's client handle. This ensures it always ticks, even if the player
+	// is standing in an unloaded chunk, unlike cPlayer::Tick. We need this because the freeze handling code must
+	// also tick in unloaded chunks.
+	if (m_IsFrozen)
+	{
+		m_FreezeCounter += 1;
+		if ((!m_IsManuallyFrozen) && (a_MyChunkIsSent))
+		{
+			cWorld::cLock Lock(*GetWorld());
+			// If the player was automatically frozen, unfreeze if the chunk the player is inside is loaded
+			Unfreeze();
+
+			// Pull the player out of any solids that might have loaded on them.
+			PREPARE_REL_AND_CHUNK(GetPosition(), *(GetParentChunk()));
+			if (RelSuccess)
+			{
+				int NewY = Rel.y;
+				if (NewY < 0)
+				{
+					NewY = 0;
+				}
+				while (NewY < cChunkDef::Height - 2)
+				{
+					// If we find a position with enough space for the player
+					if (
+						(Chunk->GetBlock(Rel.x, NewY, Rel.z) == E_BLOCK_AIR) &&
+						(Chunk->GetBlock(Rel.x, NewY + 1, Rel.z) == E_BLOCK_AIR)
+					)
+					{
+						// If the found position is not the same as the original
+						if (NewY != Rel.y)
+						{
+							SetPosition(GetPosition().x, NewY, GetPosition().z);
+							GetClientHandle()->SendPlayerPosition();
+						}
+						break;
+					}
+					++NewY;
+				}
+			}
+		}
+		else
+		{
+			// If the player was externally / manually frozen (plugin, etc.) or if the chunk isn't loaded yet:
+			// 1. Set the location to m_FrozenPosition every tick.
+			// 2. Zero out the speed every tick.
+			// 3. Send location updates every 60 ticks.
+
+			if ((m_FreezeCounter % 60 == 0) || ((m_FrozenPosition - GetPosition()).SqrLength() > 2 * 2))
```

Comment:
Oh, alright then. (zzzzzzzzzzip!)

On Fri, Apr 1, 2016 at 6:14 PM, Tiger Wang <notifications@github.com> wrote:

> In src/Entities/Player.cpp
> <https://github.com/cuberite/cuberite/pull/3097#discussion_r58229040>:
>
> > +							GetClientHandle()->SendPlayerPosition();
> > +						}
> > +						break;
> > +					}
> > +					++NewY;
> > +				}
> > +			}
> > +		}
> > +		else
> > +		{
> > +			// If the player was externally / manually frozen (plugin, etc.) or if the chunk isn't loaded yet:
> > +			// 1. Set the location to m_FrozenPosition every tick.
> > +			// 2. Zero out the speed every tick.
> > +			// 3. Send location updates every 60 ticks.
> > +
> > +			if ((m_FreezeCounter % 60 == 0) || ((m_FrozenPosition - GetPosition()).SqrLength() > 2 * 2))
>
> I am thinking that since the chunk hasn't been loaded, they would not
> suffocate. Once it is loaded, their position would be reset to a non-fatal
> one.
>
> —
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub
> <https://github.com/cuberite/cuberite/pull/3097/files/dd0ce3287ffddf3fd0dd5d7af1ebda60bd871c85#r58229040>
>